### PR TITLE
Modernize the example code blocks in the developer's guide

### LINKED ISSUE #10344: Developer guide: start advertising f strings
### Is your feature request related to a problem? Please describe.
In the developer guide, we're using old style string formatting:

```
		focusObj = api.getFocusObject()
		name = focusObj.name
		windowClassName = focusObj.windowClassName
		ui.message("class for %s window: %s" % (name, windowClassName))
```

### Describe the solution you'd like
This can be written as

```
		focusObj = api.getFocusObject()
		name = focusObj.name
		windowClassName = focusObj.windowClassName
		ui.message(f"class for {name} window: {windowClassName}")
```

### PR DESCRIPTION
**Pull request by Nael-Sayegh**
### Link to issue number:
closes #10344 

### Summary of the issue:
In the developer's guide, the code blocks have been reworked by adding "py" tags and by removing indentation spaces and on empty lines. String types of % have also been converted to f-string type.

### Description of user facing changes
In the developer guide, the user will see a new appearance of code blocks, with the string characters % replaced with string characters f, indentations by removed spaces, and py tags present at each block.

### Description of development approach
In the developer's guide, every string of type %s is manually replaced by a string of type f.

### Testing strategy:
Access the developer's guide and read the relevant sections.

### Code Review Checklist:

- [ ] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **Documentation**
  - Updated the Developer Guide to include clearer and more functional code snippets for NVDA scripts.
  - Added new methods and classes to the guide for enhancing NVDA plugin development.

### CODE REVIEW

**Review by LeonarddeR — APPROVED** [review]
Thank you!

**Review by seanbudd — APPROVED** [review]

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 1]
```diff
@@ -766,7 +766,7 @@ Once saved in the right place, either restart NVDA or choose Reload Plugins foun
     		focusObj = api.getFocusObject()
     		name = focusObj.name
     		windowClassName = focusObj.windowClassName
-    		ui.message("class for %s window: %s" % (name, windowClassName))
+    		ui.message(f"class for {name} window: {windowClassName}")
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:769]: can you please reformat this entire code block:

- remove the space indentation
- wrap the block with 3 back ticks \`\`\`py \`\`\`
- remove the start and end lines

~~~suggestion
```py
		ui.message(f"class for {name} window: {windowClassName}")
```
~~~

**Comment by CyrilleB79** [comment]
@Nael-Sayegh thanks for taking this issue.

@seanbudd, @LeonarddeR I can see that you have approved this PR. Though, it is still in draft state.

My understanding is that @Nael-Sayegh has just done the first change for now. But in #10344, the goal is to replace `%` string interpolation by f-strings for all the examples of the dev guide.

**Comment by Nael-Sayegh** [comment]
In the developer's guide, there are lines like:
ui.message("%d" % len(self.value))
Is it better to do them in the form of:
ui.message(f"{len(self.value)}")
Or:
ui.message(len(self.value))
Thank you

**Comment by CyrilleB79** [comment]
> In the developer's guide, there are lines like: ui.message("%d" % len(self.value)) Is it better to do them in the form of: ui.message(f"{len(self.value)}") Or: ui.message(len(self.value)) 

In Python console:
```
>>> import ui;ui.message(3)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "ui.pyc", line 148, in message
  File "speech\speech.pyc", line 234, in speakMessage
  File "speech\speech.pyc", line 218, in _getSpeakMessageSpeech
  File "speech\speech.pyc", line 159, in isBlank
TypeError: 'int' object is not iterable
```

So definitely:
`ui.message(f"{len(self.value)}"`

**Comment by Nael-Sayegh** [comment]
Thank for you response @CyrilleB79

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 2]
```diff
@@ -747,38 +747,36 @@ Be very careful to keep all tabs and spaces intact.
...
-    import api
-    
-    class GlobalPlugin(globalPluginHandler.GlobalPlugin):
+```py
+#Window utility scripts for NVDA
+#Developer guide example 4
+import globalPluginHandler
+from scriptHandler import script
+import ui
+import api
```
Comment (CyrilleB79) [inline projectDocs/dev/developerGuide/developerGuide.md:757]: While at it: remove spaces on blank lines to conform with NVDA's linting rules.

For example:

```suggestion

```

But there are much more blank lines that are not empty (i.e. containing spaces)

**Comment by CyrilleB79** [comment]
Also, in addition to the issue found by the Rabbit, I'd say that all the code snippet should be updated: no left margin indent, use of three backticks followed by `py`, remove spaces on blank lines, etc.)
To summarize, transform this PR in "Modernize the example code blocks in the developer's guide"; you can change the title of this PR accordingly and update the initial description.

Re the removal of "`--- start ---`" (or similar end), I agree that they cannot be part of the code snippet since they cannot be interpreted as such as Python code. Though, Though, this means that the reader of this guide should enable font name reporting or other font attribute to be able to easily localize the start and the end of the code snippet; a better solution would be to able to change voice according to text formatting (as in Jaws), but unfortunately, it's not possible in NVDA.
Alternatively, we can keep the "`--- start ---`" tags, but:
* either we put it outside of the code formatted block
* or we keep it in the code formatted block but we convert them to Python comment

@seanbudd what do you think?

**Comment by seanbudd** [comment]
@CyrilleB79 I requested removing the  start/end lines earlier in https://github.com/nvaccess/nvda/pull/16808#discussion_r1665190996

**Comment by CyrilleB79** [comment]
> @CyrilleB79 I requested removing the start/end lines earlier in [#16808 (comment)](https://github.com/nvaccess/nvda/pull/16808#discussion_r1665190996)

@seanbudd, I have seen your comment.
Though in https://github.com/nvaccess/nvda/pull/16808#issuecomment-2208707998, I have expressed my concerns about the fact that the reader may have more difficulty distinguishing the text belonging to the code examples with respect to the rest of the text of the user guide. And I have suggested possible solutions to mitigate this issue.
What do you think about this?

**Comment by Nael-Sayegh** [comment]
Thank you @CyrilleB79  I will change the title of the PR to format the entire developer's guide. I will remove the spaces and add the py tags for the different pieces of code. I am waiting for @seanbudd 's response to know how to indicate the start and end of the Python codes.

**Comment by seanbudd** [comment]
@CyrilleB79 - Is this really any more difficult to other comparably documents? I think inline code blocks in documentation are something developers would have to be used to by now. It's consistent with the rest of NVDA's more modern documentation. @Nael-Sayegh - please just go ahead and keep these removed.

**Comment by CyrilleB79** [comment]
> @CyrilleB79 - Is this really any more difficult to other comparably documents? I think inline code blocks in documentation are something developers would have to be used to by now. It's consistent with the rest of NVDA's more modern documentation. @Nael-Sayegh - please just go ahead and keep these removed.

OK @seanbudd, makes sense.
I do not know of anyone complaining of the difficulty to discriminate code vs text in NVDA's technical documentation.

**Comment by ramoncorominas** [comment]
How would the f-strings work wehen there are translatable strings? For example, if we used the following code:
```python
ui.message(_("class for %s window: %s" % (name, windowClassName)))
```

Is there a way to produce the same result but using f-strings? If not, I think the developer guide should include an example with the old style so it is clear how to include translatable strings that include variables.

**Comment by CyrilleB79** [comment]
You cannot use f-string for translatable strings.

You can do either:
* with `%` interpolation:
```
ui.message(_("class for %s window: %s") % (name, windowClassName)),
```
or
* with `.format`:
```
ui.message(_("class for {} window: {}").format(name, windowClassName))
```

Note that you can also use named parameters with `.format`.

In both cases, be careful to the position of the parenthesis, i.e. first call `_()` and then format the string. That's why the f-string cannot be used.

**Comment by Nael-Sayegh** [comment]
Hello,

I have modified the code examples to remove the start and end markers.
To comply with linting rules such as those enforced by Ruff, should we move the module imports into functions at the beginning of the code? Ruff flags this as an issue during the analysis.
What do you think?

**Comment by seanbudd** [comment]
@Nael-Sayegh - I'm not sure I understand the issue you are raising, but the imports seem fine as-is. Is ruff linting markdown now? I'm quite confused.

**Comment by CyrilleB79** [comment]
The only import that is in a function seems to be in "Example 1: An App Module that Beeps on Focus Change Events".

It's true that we usually prefer to put import statement out of functions, at the top of modules. So we may change this example with the most usual way to use import statements in NVDA.

Though are you really sure that Ruff flags the import in example 1 as an issue? This import statement is still at the beginning of the code block.

**Comment by Nael-Sayegh** [comment]
@seanbudd @CyrilleB79 's explanation is the one I was trying to convey. I haven't tested ruff on the NVDA repository, but on my module repositories, this kind of syntax is flagged as an error.
What do you think?"

**Comment by seanbudd** [comment]
I will have to review more thoroughly but I think the current examples are fine

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 3]
```diff
@@ -544,24 +548,27 @@ Be very careful to keep all tabs and spaces intact.
 Once saved in the right place, either restart NVDA or choose Reload Plugins found under Tools in the NVDA menu.
 
 From anywhere, you can now press NVDA+shift+v to have NVDA's version spoken and brailled.
+```py
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:551]:
```suggestion

```py
```

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 4]
```diff
@@ -483,22 +486,23 @@ By default, apps employing Edge WebView2 such as modern Outlook (olk.exe) are di
 ### Example 2: an app module for an app hosted by wwahost.exe {#example2}
 
 The following example is same as Notepad app module above except this is for an app hosted by wwahost.exe.
+```py
+# wwahost/test App Module for NVDA
+# Developer guide example 2
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:491]: Duplicate block

```suggestion
```

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 5]
```diff
@@ -440,21 +441,23 @@ Once saved in the correct location, either restart NVDA or choose Reload Plugins
 Finally, open Notepad and move the focus around the application; e.g. move along the menu bar, open some dialog boxes, etc.
 You should hear beeps each time the focus changes.
 Note though that if you move outside of Notepad - for instance, to Windows Explorer - you do not hear beeps.
+```py
+# Notepad App Module for NVDA
+# Developer guide example 1
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:446]: Duplicate block

```suggestion
```

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 6]
```diff
@@ -413,20 +413,21 @@ As explained above, sometimes the default way of associating an App Module with
 
 In such cases you can distribute a small global plugin along with your App Module which maps it to the executable.
 For example to map the App Module named "time_app_mod" to the "time" executable the plugin may be written as follows:
-
-    import appModuleHandler
-    import globalPluginHandler
+```py
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:416]:
```suggestion

```py
```

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 7]
```diff
@@ -413,20 +413,21 @@ As explained above, sometimes the default way of associating an App Module with
...
+		super().__init__(*args, **kwargs)
+		appModuleHandler.registerExecutableWithAppModule("time", "time_app_mod")
 
-    	def terminate(self, *args, **kwargs):
-    		super().terminate(*args, **kwargs)
-    		appModuleHandler.unregisterExecutable("time")
+	def terminate(self, *args, **kwargs):
+		super().terminate(*args, **kwargs)
+		appModuleHandler.unregisterExecutable("time")
+```
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:430]:
```suggestion
```

```

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 8]
```diff
@@ -483,22 +486,23 @@ By default, apps employing Edge WebView2 such as modern Outlook (olk.exe) are di
...
+from nvdaBuiltin.appModules.wwahost import *
 
-    --- end ---
+class AppModule(AppModule):
 
+	def event_gainFocus(self, obj, nextHandler):
+		import tones
+		tones.beep(550, 50)
+		nextHandler()
+```
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:505]:
```suggestion
```

```

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 9]
```diff
@@ -682,22 +689,20 @@ The script decorator modifies the script in such a way that it will be properly
 Furthermore, it ensures that the script is listed with the description you specify, and that it is categorised under the desired category in the input gestures dialog.
 
 In order for you to use the script decorator, you will have to import it from the scriptHandler module.
-
-    from scriptHandler import script
+```py
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:692]:
```suggestion

```py
```

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 10]
```diff
@@ -682,22 +689,20 @@ The script decorator modifies the script in such a way that it will be properly
...
-    	@script(
-    		description=_("Speaks the date and time"),
-    		category=inputCore.SCRCAT_MISC,
-    		gestures=["kb:NVDA+shift+t", "kb:NVDA+alt+r"]
-    	)
-    	def script_sayDateTime(self, gesture):
-
-    --- end ---
-
+```py
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:698]:
```suggestion

```py
```

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 11]
```diff
@@ -682,22 +689,20 @@ The script decorator modifies the script in such a way that it will be properly
...
+	@script(
+		description=_("Speaks the date and time"),
+		category=inputCore.SCRCAT_MISC,
+		gestures=["kb:NVDA+shift+t", "kb:NVDA+alt+r"]
+	)
+	def script_sayDateTime(self, gesture):
+```
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:705]:
```suggestion
```

```

File: projectDocs/dev/developerGuide/developerGuide.md [hunk 12]
```diff
@@ -747,38 +752,37 @@ Be very careful to keep all tabs and spaces intact.
...
+	@script(
+		description=_("Announces the window control ID of the current focus object"),
+		gesture="kb:NVDA+rightArrow"
+	)
+	def script_announceWindowControlID(self, gesture):
+		focusObj = api.getFocusObject()
+		name = focusObj.name
+		windowControlID = focusObj.windowControlID
+		ui.message(f"Control ID for {name} window: {windowControlID}")
+```
```
Comment (seanbudd) [inline projectDocs/dev/developerGuide/developerGuide.md:785]:
```suggestion
```

```

**Review by seanbudd — CHANGES_REQUESTED** [review]
Please ensure there are new lines before and after the start of code blocks.

There are many code blocks here with duplicated fragments.

Please perform a thorough review - I encourage reviewing with and without whitespace viewing

**Comment by Nael-Sayegh** [comment]
@seanbudd I have made the modifications, what do you think?

**Comment by Nael-Sayegh** [comment]
@seanbudd Do these modifications require an entry in the change log?

**Review by seanbudd — APPROVED** [review]

**Comment by seanbudd** [comment]
Thanks @Nael-Sayegh - no need for a changelog entry

**Pull request merged**